### ISSUE TITLE
[java] UnnecessaryCast: False positive for anonymous class in Stream.map() with toList()

### ISSUE THREAD

**Issue by kna** [pmd/pmd#6409]
**Affects PMD Version:** 
7.20.0

**Rule:**
[UnnecessaryCast](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessarycast)

**Description:**
PMD reports `UnnecessaryCast` for a cast that is actually **required** for successful compilation when returning an anonymous class implementing an interface from a `Stream.map()` lambda expression followed by `.toList()`.

The cast is necessary due to how Java's type inference handles anonymous classes - without it, the code fails to compile with a type incompatibility error.

**Code Sample demonstrating the issue:**

```java
package org.example;

import java.util.List;

public class Sample {
    public List<Foo> calc(final List<Integer> list) {
        return list.stream().map(i -> (Foo) new Foo(){}).toList();
    }
    interface Foo{}
}
```

**Expected outcome:**

PMD reports a violation at line 7, but that's wrong. That's a false positive.

If you delete the cast, you fail to compile it whith this error:
```
error: incompatible types: List<<anonymous Foo>> cannot be converted to List<Foo>
        return list.stream().map(i -> new Foo(){}).toList();
```

**Running PMD through:**
Gradle
Java 17

**Comment by zbynek** [comment]
You don't need the cast if you use

```
list.stream().<Foo>map(i -> new Foo(){}).toList();
```

Not sure if PMD should force you to do that though.
